Provide Thumbor 7 / Python 3 compatibility

relthyg opened the pull request:
This PR provides Thumbor 7 / Python 3 compatibility.

Co-authored-by: mpdude <mp@webfactory.de>

relthyg committed:
Provide Thumbor 7 / Python 3 compatibility

mpdude commented:
@heynemann can you confirm this looks about right?

heynemann commented:
I don't know much about the PDF library, but seems to be right :)

mpdude commented:
I don’t know much about Python, but it seems to run 😂

@intellisense WDYT?

intellisense commented:
@relthyg Hi! Thanks for the PR. Been a long time since I have looked back at this repository. I will play with the changes over the weekend, and If all looks good, I will create a new release. Thank you!

relthyg commented:
Great, thank you. Let me know if something needs to be improved. Have a nice weekend!

mpdude commented:
There may still be some problems with the URL signing (HMAC) lurking here – works in dev with unsafe URLs, but not in prod with secrets enabled. Admittedly, the PDF origin URLS also include a query string, but were fully URL-encoded.

relthyg committed:
Decode hash parameter of `preview_url`

`RequestParser.path_to_parameters()` returns encoded parameters, but we need a decoded hash in order to pass the url validation in `ImagingHandler.check_image()` later on.

relthyg committed:
Encode unquoted `url_signature` before validation

`url_signature = self.context.request.hash` is an unencoded string, but `UrlSigner.signature()` from `libthumbor` delivers an encoded string. So we must encode our `url_signature` in order to be able to compare the two.

relthyg commented:
@intellisense URL signing works now for safe URLs too.

intellisense commented on tc_pdf/handlers/pdf.py:
```diff
@@ -51,10 +50,9 @@ def regex(cls, has_unsafe_or_hash=True):
+    async def get(self, **kwargs):
         # check if request is valid
-        yield gen.maybe_future(self.check_pdf(kwargs.copy()))
+        await self.check_pdf(kwargs.copy())
```
There is a minor problem here. If `check_pdf` finishes the request early because of validation we need to return early here.
```suggestion
        await self.check_pdf(kwargs.copy())
        # return early if at this point the request isn't valid
        status = self.get_status()
        if status != 200:
            return
```

Otherwise if an unsafe URL is provided an unsafe URLs are not allowed then we get an error:
```
2023-05-15 19:49:08 thumbor:ERROR ERROR: Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/tornado/web.py", line 1786, in _execute
    result = await result
  File "/app/tc_pdf/handlers/pdf.py", line 103, in get
    return await super(PDFHandler, self).get(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/thumbor/handlers/imaging.py", line 119, in get
    return await self.check_image(kw)
  File "/usr/local/lib/python3.9/site-packages/thumbor/handlers/imaging.py", line 63, in check_image
    self._error(
  File "/usr/local/lib/python3.9/site-packages/thumbor/handlers/__init__.py", line 116, in _error
    self.finish()
  File "/usr/local/lib/python3.9/site-packages/tornado/web.py", line 1201, in finish
    raise RuntimeError("finish() called twice")
RuntimeError: finish() called twice
```

relthyg replied:
@intellisense Same here, this is indeed a problem, but it's not introduced by the changes that we made to provide Thumbor 7 / Python 3 compatibility, is it? Are yo sure you want to fix this in this PR or maybe open another?

intellisense replied:
Agreed, however there was a change that I forgot to comment now requested [here](https://github.com/intellisense/tc_pdf/pull/3#discussion_r1195226983). If you are comfortable making this change lets do it otherwise I can do it separately. Thanks!

relthyg replied:
I'm fine with it, it's your project ;-)

intellisense commented on tc_pdf/handlers/pdf.py:
```diff
@@ -115,11 +116,11 @@ def create_preview(self, url, resolution=200):
+                            return out_io.getvalue()
 
                 except WandException as e:
                     logger.exception('[PDFHander.create_preview] %s', e)
```
```suggestion
                    logger.exception('[PDFHandler.create_preview] %s', e)
```

relthyg replied:
I'm not sure if this change is in the scope of this PR, because the typo wasn't introduced by it, but anyways it's better to fix it.

intellisense reviewed (changes requested):
@relthyg Thank you for the PR and contributing to this 👏. Just a couple of minor issues and after that it is good to be merged.

There are other improvements that could be made but I will do that separately. Thanks!

relthyg committed:
Fix typo

Co-authored-by: Aamir Rind <aamir.adnan.rind@gmail.com>

intellisense commented on tc_pdf/handlers/pdf.py:
```diff
@@ -63,21 +61,25 @@ def get(self, **kwargs):
-        kwargs['hash'] = RequestParser.path_to_parameters(preview_url)['hash']
+        parameters = RequestParser.path_to_parameters(preview_url)
+        for parameter in parameters:
+            if parameter[0] == 'hash':
```
```suggestion
            if parameter[0] == 'hash' and parameter[1]:
```

Otherwise if unsafe URLs are allowed and an unsafe URL is provided there is an error:
```
2023-05-15 19:10:41 thumbor:ERROR ERROR: Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/tornado/web.py", line 1786, in _execute
    result = await result
  File "/app/tc_pdf/handlers/pdf.py", line 82, in get
    kwargs['hash'] = parameter[1].decode()
AttributeError: 'NoneType' object has no attribute 'decode'
```

relthyg committed:
Check whether value for `hash` is actually provided

Co-authored-by: Aamir Rind <aamir.adnan.rind@gmail.com>

relthyg committed:
Fail fast if `check_pdf()` receives non-200 status code

Co-authored-by: Aamir Rind <aamir.adnan.rind@gmail.com>

relthyg commented:
Would you mind tagging a new release?

intellisense commented:
@relthyg Available now with version `1.0.0`. Thanks!
